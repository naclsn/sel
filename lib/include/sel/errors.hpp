#ifndef SEL_ERROR_HPP
#define SEL_ERROR_HPP

/**
 * The various types of errors that can occure:
 * BaseError, ParseError, TypeError, RuntimeError.
 * NIYError is kept for now as a wildcard
 */

#include <stdexcept>
#include <sstream>
#include <string>

#include "types.hpp"
#include "engine.hpp"

namespace sel {

  struct BaseError : std::runtime_error {
    BaseError(std::string const& msg)
      : std::runtime_error(msg)
    { }
  };

  struct NIYError : BaseError {
    NIYError(std::string const&  missing)
      : BaseError("not implemented yet: " + missing)
    { }
  };

  struct ParseError : BaseError {
    size_t start, span;
    ParseError(std::string const& msg, size_t start, size_t span)
      : BaseError(msg)
      , start(start)
      , span(span)
    { }
  };

  struct TypeError : BaseError {
    // size_t start, span;
    TypeError(std::string const& msg) // range..
      : BaseError(msg)
    { }
    TypeError(Type const& from, Type const& to) // range..
      : TypeError(bidoof(from, to))
    { }
  private:
    inline static std::string bidoof(Type const& from, Type const& to) {
      std::ostringstream oss;
      return (oss << "cannot coerse from " << from << " to " << to, oss.str());
    }
  };

  struct RuntimeError : BaseError {
    RuntimeError(std::string const& msg)
      : BaseError(msg)
    { }
  };

} // namespace sel

#endif // SEL_ERROR_HPP
