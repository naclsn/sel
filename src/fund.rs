//! fundamental operations

use std::fmt::{Display, Formatter, Result as FmtResult};
use std::rc::Rc;

use crate::types::Type;

/// macro to generate the fn generating the type, internal to `mkmkty!`
macro_rules! mkmktyty {
    ($st:ident, Num) => {
        Type::number()
    };

    ($st:ident, Str+ $f:literal $(& $both:literal)? $(| $either:literal)?) => {
        {
            let f = mkmktyty!($st, + $f $(& $both)? $(| $either)?);
            Type::bytes(f)
        }
    };
    ($st:ident, Str) => {
        mkmktyty!($st, Str+ 0)
    };

    ($st:ident, $n:ident) => {
        $st.$n.clone()
    };

    ($st:ident, [$($i:tt)+]+ $f:literal $(& $both:literal)? $(| $either:literal)?) => {
        {
            let f = mkmktyty!($st, + $f $(& $both)? $(| $either)?);
            let i = mkmktyty!($st, $($i)+);
            Type::list(f, i)
        }
    };
    ($st:ident, [$($i:tt)+]) => {
        mkmktyty!($st, [$($i)+]+ 0)
    };

    ($st:ident, (($($a:tt)+), $($b:tt)+)) => {
        {
            let f = mkmktyty!($st, ($($a)+));
            let s = mkmktyty!($st, $($b)+);
            Type::pair(f, s)
        }
    };
    ($st:ident, ($a:tt, $($b:tt)+)) => {
        mkmktyty!($st, (($a), $($b)+))
    };
    ($st:ident, ($a:tt+ $f:literal $(& $both:literal)? $(| $either:literal)?, $($b:tt)+)) => {
        mkmktyty!($st, (($a+ $f $(& $both)? $(| $either)?), $($b)+))
    };

    ($st:ident, ($($t:tt)+)) => {
        mkmktyty!($st, $($t)+)
    };

    ($st:ident, $p:tt -> $($r:tt)+) => {
        {
            let p = mkmktyty!($st, $p);
            let r = mkmktyty!($st, $($r)+);
            Type::func(p, r)
        }
    };
    ($st:ident, $p:tt+ $f:literal $(& $both:literal)? $(| $either:literal)? -> $($r:tt)+) => {
        mkmktyty!($st, ($p+ $f $(& $both)? $(| $either)?) -> $($r)+)
    };

    ($st:ident, + $n:literal) => {
        false //$st._fin[$n]
    };
    ($st:ident, + $l:literal&$r:literal) => {
        false //$st.types.both($st._fin[$l], $st._fin[$r])
    };
    ($st:ident, + $l:literal|$r:literal) => {
        false //$st.types.either($st._fin[$l], $st._fin[$r])
    };
}

/// macro to generate the fn generating the type: (nb_infinites, names; type)
macro_rules! mkmkty {
    ($f:tt$(, $n:ident)*; $($ty:tt)+) => {
        || {
            struct State {
                //_fin: [Boundedness; $f+1],
                $($n: Rc<Type>,)*
            }
            let _state = State {
                //_fin: mkmkty!(@ types $f),
                $($n: Type::named(stringify!($n).into()),)*
            };
            mkmktyty!(_state, $($ty)+)
        }
    };

    //(@ $types:ident 0) => { [$types.finite(true)] };
    //(@ $types:ident 1) => { [$types.finite(true), $types.finite(false)] };
    //(@ $types:ident 2) => { [$types.finite(true), $types.finite(false), $types.finite(false)] };
    //(@ $types:ident 3) => { [$types.finite(true), $types.finite(false), $types.finite(false), $types.finite(false)] };
}

macro_rules! make {
    ($($(#[$desc:meta])+ $var:ident($name:literal) :: $mkty:expr),*$(,)?) => {
        /// Fundamental operations that cannot be implemented in the language itself.
        ///
        /// This does not mean everything else can be done within the language;
        /// eg OS-level interractions are to be implemented somewhere else. Likewise,
        /// not all maths may be accessible with that limited of a set (idk really).
        ///
        /// It basically represent what an interpreter will in any way need to implement.
        /// Technically `pipe`, `tonum` and `tostr` _could_ be implemented in the prelude
        /// but are not for consistency with other syntax-level functions.
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
        pub enum Fund { $($(#[$desc])+ $var,)* }

        impl Display for Fund {
            fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
                write!(f, "{}", match self { $(Self::$var => $name,)* })
            }
        }

        impl Fund {
            pub fn try_from_name(name: &str) -> Option<Self> {
                match name { $($name => Some(Self::$var),)* _ => None }
            }

            pub fn make_type(&self) -> Rc<Type> {
                match self { $(Self::$var => ($mkty as fn() -> Rc<Type>)(),)* }
            }

            pub fn desc(&self) -> &'static [&'static str] {
                match self { $(Self::$var => &[$(stringify!($desc),)+],)+ }
            }
        }
    };
}

make! {
    /// the input
    Stream("-")                  :: mkmkty!(1          ; Str+1                   ),
    /// the type-restricting function
    Astypeof("astypeof")         :: mkmkty!(0, a       ; a -> a -> a             ),
    /// make a list with the head element first, then the rest as tail;
    /// 'snoc lst val' is equivalent to the syntax '{val,, lst}'
    // we use snoc and not cons as it makes better error messages (because it binds the list first)
    Snoc("snoc")                 :: mkmkty!(1, a       ; [a]+1 -> a -> [a]+1     ),
    /// panics (runtime abort with message)
    Panic("panic")               :: mkmkty!(0, a       ; Str -> a                ),
    /// pipe two functions;
    /// 'pipe one two' is equivalent to the syntax 'one, two' ie 'two(one(..))'
    /// (see also 'compose')
    Pipe("pipe")                 :: mkmkty!(0, a, b, c ; (a->b) -> (b->c) -> a->c),
    /// convert a string into number;
    /// accept an infinite string for convenience but stop on the first invalid byte
    Tonum("tonum")               :: mkmkty!(1          ; Str+1 -> Num            ),
    /// convert a number into string
    Tostr("tostr")               :: mkmkty!(0          ; Num -> Str              ),

    /// make a list of numbers with the 8 bits bytes
    Bytes("bytes")               :: mkmkty!(1          ; Str+1 -> [Num]+1        ),
    /// make a list of numbers with the 32 bits codepoints
    Codepoints("codepoints")     :: mkmkty!(1          ; Str+1 -> [Num]+1        ),
    /// make a list of strings with the potentially multi-codepoints graphemes
    Graphemes("graphemes")       :: mkmkty!(1          ; Str+1 -> [Str]+1        ),
    /// make a string from the 8 bits bytes
    Unbytes("unbytes")           :: mkmkty!(1          ; [Num]+1 -> Str+1        ),
    /// make a string from the 32 bits codepoints
    Uncodepoints("uncodepoints") :: mkmkty!(1          ; [Num]+1 -> Str+1        ),
    /// make a string from the potentially multi-codepoints graphemes
    Ungraphemes("ungraphemes")   :: mkmkty!(1          ; [Str]+1 -> Str+1        ),

    /// add two numbers
    Add("add")                   :: mkmkty!(0          ; Num -> Num -> Num       ),
    /// invert a number
    Invert("invert")             :: mkmkty!(0          ; Num -> Num              ),
    /// multiply two numbers
    Mul("mul")                   :: mkmkty!(0          ; Num -> Num -> Num       ),
    /// negate a number
    Negate("negate")             :: mkmkty!(0          ; Num -> Num              ),
    /// sign of a number, in {-1, 0, 1}
    Signum("signum")             :: mkmkty!(0          ; Num -> Num              ),
    /// truncation (rounding toward zero)
    Trunc("trunc")               :: mkmkty!(0          ; Num -> Num              ),

    /// the inverse sine function
    Asin("asin")                 :: mkmkty!(0          ; Num -> Num              ),
    /// the exponent function
    Exp("exp")                   :: mkmkty!(0          ; Num -> Num              ),
    /// the natural logarithm function
    Log("log")                   :: mkmkty!(0          ; Num -> Num              ),
    /// the sine function
    Sin("sin")                   :: mkmkty!(0          ; Num -> Num              ),
}
