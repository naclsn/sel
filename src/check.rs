//! checking and lowering into an AST

use std::collections::HashMap;
use std::rc::Rc;

use crate::errors::{self, Error, ErrorKind};
use crate::fund::Fund;
use crate::module::{Function, Location, Module, ModuleRegistry};
use crate::parse::{Apply, ApplyBase, Located, Pattern, Script, Value};
use crate::types::Type;

#[derive(Debug, Clone)]
pub struct Tree {
    pub ty: Rc<Type>,
    //pub loc: Location, // XXX(wip)?
    pub val: TreeVal,
}

#[derive(Debug, Clone)]
pub enum TreeVal {
    Number(f64),
    Bytes(Box<[u8]>),
    Word(String, Refers),
    List, // represents an empty list, as a cons-sequence sentinel (ie nil)
    Pair(Box<Tree>, Box<Tree>),
    Apply(Box<Tree>, Vec<Tree>), // base is 'Word' or 'Binding', or the error list isn't empty
    Binding(Pattern, Box<Tree>, Option<Box<Tree>>), // Some(.) iff `pat.is_refutable()`
                                 //Fundamental(Fund),
}

#[derive(Debug, Clone)]
pub enum Refers {
    Fundamental(Fund),              // eg cons, panic, bytes...
    Binding(Location),              // from an enclosing let binding
    Defined(Rc<Function>),          // from within this file (def)
    File(Rc<Module>, Rc<Function>), // from external (user) file
    //XXX(wip) Builtin(String),   // eg from prelude // maybe 'll just use module named "<prelude>"
    Missing, // it's shoved in the nearest enclosing scope
}

pub struct Checker<'check> {
    registry: &'check ModuleRegistry,
    module: &'check Module,
    scopes: Vec<HashMap<String, (Location, Rc<Type>)>>, // always at least one
    errors: Vec<Error>,
}

impl<'check> Checker<'check> {
    pub fn new(module: &'check Module, registry: &'check ModuleRegistry) -> Self {
        Self {
            registry,
            module,
            scopes: vec![Default::default()], // top-level fallback scope
            errors: Vec::new(),
        }
    }

    pub fn errors(&self) -> &[Error] {
        &self.errors
    }

    pub fn boxed_errors(self) -> Box<[Error]> {
        self.errors.into()
    }

    /// lookup order is always:
    /// - fundamentals
    /// - bindings (closest to furthest)
    /// - defineds (latest to earliest in file)
    /// - uses (same order)
    /// - missing
    ///
    /// when None, likely want to call `missing()`
    /// when ToBeDefined, likely want to call `define()`
    fn lookup(&mut self, loc: &Location, name: &str) -> (Rc<Type>, Refers) {
        if let Some(fund) = Fund::try_from_name(name) {
            return (fund.make_type(), Refers::Fundamental(fund));
        }

        if let Some((old_loc, ty)) = self.scopes.iter().rev().find_map(|it| it.get(name)) {
            return (ty.clone(), Refers::Binding(old_loc.clone()));
        }

        if let Some(def) = self.module.retrieve_function(name, self.registry) {
            return (Type::deep_clone(&def.ast.ty), Refers::Defined(def));
        }

        if let Some((prefix, name)) = name.split_once('-') {
            match self.module.retrieve_module(prefix, self.registry) {
                None => {
                    todo!("specialize error: rather than just 'missing', have 'no use :: {prefix}'")
                }
                Some(Err(err)) => todo!("cannot load {prefix}: {err:?}"),
                Some(Ok(module)) => {
                    if let Some(def) = module.retrieve_function(name, self.registry) {
                        return (Type::deep_clone(&def.ast.ty), Refers::File(module, def));
                    } else {
                        todo!("specialize error: rather than just 'missing', have 'no {name} in {prefix}'");
                    }
                }
            }
        }

        self.errors.push(errors::unknown_name(
            loc.clone(),
            name.into(),
            &Type::named("TODO: todou (unkown_name type after evrtynthsgslkfdjs)".into()),
        ));
        let ty = Type::named(name.into());
        self.scopes
            .last_mut()
            .expect("should always have at least top-level fallback scope")
            .insert(name.into(), (loc.clone(), ty.clone()));
        // TODO push something in scope, so we dont get here again from this case
        (ty, Refers::Missing)
    }

    fn apply(&mut self, loc_func: Location, func: Tree, loc_arg: Location, arg: Tree) -> Tree {
        let ty = match &*func.ty {
            Type::Func(par, ret) => {
                if let Err(err) = Type::concretize(loc_func, par, loc_arg, &arg.ty) {
                    self.errors.push(err);
                }
                ret.clone()
            }
            sad => {
                self.errors
                    .push(errors::not_function(loc_func, sad, &func.val));
                Type::named(format!("ret"))
            }
        };

        let val = match func.val {
            TreeVal::Apply(base, mut args) => {
                args.push(arg);
                TreeVal::Apply(base, args)
            }
            TreeVal::Word(_, _) | TreeVal::Binding(_, _, _) => {
                TreeVal::Apply(Box::new(func), vec![arg])
            }
            TreeVal::Number(_) | TreeVal::Bytes(_) | TreeVal::List | TreeVal::Pair(_, _) => {
                TreeVal::Apply(Box::new(func), vec![arg])
            }
        };

        Tree { ty, val }
    }

    pub fn check_script(&mut self, script: &Script) -> Tree {
        let val = self.check_apply(&script.head);
        let loc = script.head.loc();

        if matches!(*val.ty, Type::Func(_, _)) {
            // [f, g, h] -> [pipe f g, h] -> pipe [pipe f g] h
            script
                .tail
                .iter()
                .fold((loc, val), |(loc_prev, acc), (loc_comma, cur)| {
                    let then = self.check_apply(cur);
                    let loc_cur = cur.loc();
                    let pipe = Tree {
                        ty: Fund::Pipe.make_type(),
                        val: TreeVal::Word(Fund::Pipe.to_string(), Refers::Fundamental(Fund::Pipe)),
                    };
                    let part = self.apply(loc_comma.clone(), pipe, loc_prev, acc);
                    let res = self.apply(loc_comma.clone(), part, loc_cur.clone(), then);
                    (loc_cur, res)
                })
                .1
        } else {
            // [x, f, g, h] -> h [g [f x]]
            script
                .tail
                .iter()
                .fold((loc, val), |(loc_prev, acc), (_loc_comma, cur)| {
                    let func = self.check_apply(cur);
                    let loc_cur = cur.loc();
                    (loc_cur.clone(), self.apply(loc_cur, func, loc_prev, acc))
                })
                .1
        }
    }

    fn check_pattern_type(
        &mut self,
        pat: &Pattern,
        names: &mut HashMap<String, (Location, Rc<Type>)>,
    ) -> Rc<Type> {
        match pat {
            Pattern::Number { .. } => Type::number(),

            Pattern::Bytes { .. } => {
                let fin = true;
                Type::bytes(fin)
            }

            Pattern::Word { loc, word } => {
                let ty = Type::named(word.clone());

                let loc = loc.clone();
                use std::collections::hash_map::Entry::*;
                match names.entry(word.clone()) {
                    Occupied(old) => {
                        let (loc_already, ty_old) = old.get();
                        self.errors.push(errors::already_declared(
                            word.clone(),
                            loc_already.clone(),
                            &*ty_old,
                            loc,
                        ));
                        ty_old.clone()
                    }
                    Vacant(niw) => niw.insert((loc, ty)).1.clone(),
                }
            }

            Pattern::List {
                loc_open: _,
                items,
                rest,
                loc_close: _,
            } => {
                let items: Box<_> = items
                    .iter()
                    .map(|it| self.check_pattern_type(it, names))
                    .collect();

                // i kept the commented code below that refers to check_list_content;
                // i just rewrote Value::List handling in check_value in a way
                // that's more regular: it's always checked as if {...,, {}} ie it
                // always relies on `cons` which will dictate the behavior uniformly
                // however i'm left with that one now which idk how to then

                //let ty: Rc<Type>;// = todo!("");
                //let has = self.check_list_content(items);
                //let fin = self.types.finite(rest.is_none());
                //let ty = self.types.list(fin, has);

                let ty = Type::list(
                    rest.is_none(),
                    todo!("ty of Pattern::List\nitem types: {items:?}"),
                );

                if let Some((loc_comma, loc, word)) = rest {
                    let loc = loc.clone();
                    use std::collections::hash_map::Entry::*;
                    match names.entry(word.clone()) {
                        Occupied(old) => {
                            let (loc_already, ty_old) = old.get();
                            self.errors.push(errors::context_extra_comma_makes_rest(
                                loc_comma.clone(),
                                errors::already_declared(
                                    word.clone(),
                                    loc_already.clone(),
                                    &*ty_old,
                                    loc,
                                ),
                            ));
                        }
                        Vacant(niw) => _ = niw.insert((loc, ty)),
                    }
                }
                ty
            }

            Pattern::Pair {
                fst,
                loc_equal: _,
                snd,
            } => {
                let fst = self.check_pattern_type(fst, names);
                let snd = self.check_pattern_type(snd, names);
                Type::pair(fst, snd)
            }
        }
    }

    fn check_binding_br(
        &mut self,
        pat: &Pattern,
        res: &Value,
        alt: Option<&Value>,
    ) -> (Rc<Type>, Tree, Option<Tree>) {
        let mut names = HashMap::new();
        let param = self.check_pattern_type(pat, &mut names);
        self.scopes.push(names);

        let res = self.check_value(res);

        let alt = alt.map(|alt| {
            assert!(pat.is_refutable());
            let alt = self.check_value(alt);
            if let Result::<(), ErrorKind>::Err(err) = todo!("Type::harmonize(res.ty, alt.ty)") {
                //self.errors.push(Error(
                //    todo!("loc for {err:?} (from harmonize res&alt)"),
                //    err,
                //    //ErrorKind::ContextCaused {
                //    //    error: Box::new(Error(result.loc.clone(), e)),
                //    //    because: ErrorContext::LetFallbackTypeMismatch {
                //    //        result_type: snapshot.frozen(res_ty),
                //    //        fallback_type: snapshot.frozen(fallback.ty),
                //    //    },
                //    //},
                //));
            }
            alt
        });

        self.scopes.pop();
        (Type::func(param, res.ty.clone()), res, alt)
    }

    pub fn check_apply(&mut self, apply: &Apply) -> Tree {
        let base = match &apply.base {
            ApplyBase::Binding {
                loc_let: _,
                pat,
                res,
                alt,
            } => {
                let (ty, res, alt) = self.check_binding_br(pat, &res, alt.as_deref());
                let val = TreeVal::Binding(pat.clone(), Box::new(res), alt.map(Box::new));
                Tree { ty, val }
            }
            ApplyBase::Value(val) => self.check_value(val),
        };

        let loc_base = apply.loc();
        apply.args.iter().fold(base, |acc, cur| {
            let arg = self.check_value(cur);
            self.apply(loc_base.clone(), acc, cur.loc(), arg)
        })
    }

    // not used, but kept for now because of the notes;
    // see in check_pattern_type's Pattern::List branch too
    fn check_list_content(&mut self, items: impl IntoIterator<Item = Rc<Type>>) -> Rc<Type> {
        let ty = Type::named("item".into());
        for it in items {
            _ = it;
            // so, unless I'm off, I think the way I did rest part for literal lists
            // will not allow eg {inf, fin,, {}} because it would go:
            //      cons fin {} :: fin
            //      cons inf fin :: type error!
            // looking again, I think that was already the case in previous implementation,
            // so now I'm wondering if I should also disallow when no rest for both
            // consistency and removing hunks from `types` (and almost idealy whole `types.rs`)
            //
            // but does that makes sens? isn't that annoying that it won't be possible to eg
            //      {:hello:, -}, unwords
            // but say in this specific case, maybe the miss is to say literal bytes is inf..?
            //if let Err(err) = Type::harmonize(ty, it) {
            //    self.errors
            //        .push(todo!("error::list_type_mismatch(..) for {err:?}"));
            //}
            // constructing always with `cons` seems like the least sane but most regular approach
            // (insane because that' a de-optimisation which will /have/ to be corrected further down the line)
            //
            // cons <fin> [<fin>] :: [<fin>] -- free
            // cons <inf> [<inf>] :: [<inf>] -- free
            // cons <fin> [<inf>] :: [<inf>] -- can never work through type sys only
            // cons <inf> [<fin>] :: [<inf>] -- works if (no-op) coersion [<fin>] -> [<inf>]
        }
        ty
    }

    pub fn check_value(&mut self, value: &Value) -> Tree {
        match value {
            Value::Number { loc: _, number } => Tree {
                ty: Type::number(),
                val: TreeVal::Number(*number),
            },

            Value::Bytes { loc: _, bytes } => Tree {
                ty: {
                    let fin = true;
                    Type::bytes(fin)
                },
                val: TreeVal::Bytes(bytes.clone()),
            },

            Value::Word { loc, word } => {
                let (ty, prov) = self.lookup(loc, word);
                let val = TreeVal::Word(word.clone(), prov);
                Tree { ty, val }
            }

            Value::Subscr {
                loc_open: _,
                subscr,
                loc_close: _,
            } => self.check_script(subscr),

            Value::List {
                loc_open,
                items,
                rest,
                loc_close: _,
            } => {
                let items: Vec<_> = items.iter().map(|it| (it, self.check_apply(it))).collect();

                let (loc_rest, rest) = if let Some((_loc_comma, rest)) = rest {
                    let mut maybe_list = self.check_apply(rest);
                    // if `ty` is named, noone here is referring to it yet so things will get messy
                    // once applying `snoc` wants to upgrade referring types to now referring to
                    // the list; instead ensure it's made a list at this point
                    // (any other type will be properly reported, see below)
                    // XXX: no, that's not enough as other things can refer to types: bindings
                    if matches!(&*maybe_list.ty, Type::Named(_, _)) {
                        maybe_list.ty = Type::list(false, maybe_list.ty);
                    }
                    (rest.loc(), maybe_list)
                } else {
                    // empty list, as if {...,, {}}
                    let nil = Tree {
                        ty: Type::list(true, Type::named("item".into())),
                        val: TreeVal::List,
                    };
                    // with Type::list of nil, first apply of snoc cannot fail (:: [a] -> a -> [a])
                    (Location(Default::default(), 0..0), nil)
                };

                // cons(items[0], .... cons(items[n-1], cons(items[n], rest)) .... )
                // note: reverse error order -- XXX: may not want that
                // TODO: if error anyways might want to grab some context to see if we can
                //       find the type that it should be; note: this might just be insane..
                //       at least a ContextCaused "lists are checked from the end" and group
                //       all type err in
                items.into_iter().rfold(rest, |acc, (cur, val)| {
                    let snoc = Tree {
                        ty: Fund::Snoc.make_type(),
                        val: TreeVal::Word(Fund::Snoc.to_string(), Refers::Fundamental(Fund::Snoc)),
                    };
                    // side-note: loc_rest is only -maybe- require for the first iteration,
                    // when `acc` (=`rest`) may not be a list; the rest of the time clone is usl
                    let part = self.apply(loc_open.clone(), snoc, loc_rest.clone(), acc);
                    self.apply(loc_open.clone(), part, cur.loc(), val)
                })
            }

            Value::Pair {
                fst,
                loc_equal: _,
                snd,
            } => {
                let fst = Box::new(self.check_value(fst));
                let snd = Box::new(self.check_value(snd));
                Tree {
                    ty: Type::pair(fst.ty.clone(), snd.ty.clone()),
                    val: TreeVal::Pair(fst, snd),
                }
            }
        }
    }
}

#[test]
fn test() {
    use insta::assert_debug_snapshot;

    fn t(script: &[u8]) -> (Tree, Box<[Error]>) {
        let registry = ModuleRegistry::default();
        let module = registry.load_bytes("<test>", script.iter().copied());
        let function = module.retrieve(&registry).unwrap();
        (function.ast, function.errors)
    }

    assert_debug_snapshot!(t(b"add 1 2"));
    assert_debug_snapshot!(t(b"{1, 2, 3}"));
    //assert_debug_snapshot!(t(b"{1, 2, 3, :soleil:}"));
}
