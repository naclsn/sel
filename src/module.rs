//! file-level scoping and resolving

use std::cell::RefCell;
use std::collections::HashMap;
use std::io::Result as IoResult;
use std::ops::Range;
use std::rc::Rc;

use crate::check::{Checker, Tree};
use crate::error::Error;
use crate::parse::{Parser, Top};

#[derive(PartialEq, Debug, Clone)]
pub struct Location(pub String, pub Range<usize>);

#[derive(Debug)]
pub struct Function {
    pub ast: Tree,
    pub errors: Box<[Error]>,
}

#[derive(Debug)]
pub struct Module {
    pub path: String,
    pub bytes: Box<[u8]>,
    line_map: Box<[Range<usize>]>,
    top: Top,
    pub errors: Box<[Error]>,
    functions: RefCell<HashMap<String, Rc<Function>>>,
}

#[derive(Default)]
pub struct ModuleRegistry {
    modules: RefCell<HashMap<String, Rc<Module>>>,
}

impl ModuleRegistry {
    // i dont like the use of 'load' with immut args, might rename to 'require' or idk but donnn
    // mater much anyways

    /// load `bytes` as a module's source and associate them with `path`
    ///
    /// `path` isn't altered/normalized in any way
    ///
    /// parsing in and by itself is not recursive, this does not hold on to a borrow
    /// during parsing (until insertion into the hash map)
    pub fn load_bytes(&self, path: &str, bytes: impl IntoIterator<Item = u8>) -> Rc<Module> {
        let bytes: Box<[u8]> = bytes.into_iter().collect();
        let line_map = Module::compute_line_map(&bytes);

        let mut parser = Parser::new(&path, bytes.iter().copied());
        let top = parser.parse_top();
        let errors = parser.boxed_errors();
        let module = Module {
            path: path.into(),
            bytes,
            line_map,
            top,
            errors,
            functions: Default::default(),
        };

        self.modules
            .borrow_mut()
            .entry(path.into())
            .insert_entry(module.into())
            .get()
            .clone()
    }

    /// try to load a module from `path` (ie fs as of now)
    ///
    /// `path` may be altered/normalized
    ///
    /// parsing in and by itself is not recursive, this does not hold on to a borrow
    /// even if it hadn't been parsed yet (until insertion into the hash map)
    pub fn load(&self, path: &str) -> IoResult<Rc<Module>> {
        let has = self.modules.borrow().get(path).cloned();
        if let Some(found) = has {
            return Ok(found);
        }

        Ok(self.load_bytes(path, std::fs::read(path)?))
    }
}

impl Module {
    /// the top level
    ///
    /// checking may be recursive if the function refers to 'def's
    /// within this module or a 'use'd module; registry's load may
    /// be invoked as needed
    pub fn retrieve(&self, registry: &ModuleRegistry) -> Option<Function> {
        let mut checker = Checker::new(self, registry);
        Some(Function {
            ast: checker.check_script(self.top.script.as_ref()?),
            errors: checker.boxed_errors(),
        })
    }

    /// search in 'use's
    ///
    /// functions are checked lazily, however this does not hold on to a borrow
    /// even if it hadn't been checked yet (until insertion into the hash map)
    pub fn retrieve_function(&self, name: &str, registry: &ModuleRegistry) -> Option<Rc<Function>> {
        let has = self.functions.borrow().get(name).cloned();
        if has.is_some() {
            return has;
        }

        let apply = &self.top.defs.iter().rev().find(|d| name == d.name)?.to;
        let mut checker = Checker::new(self, registry);
        let function = Function {
            ast: checker.check_apply(apply),
            errors: checker.boxed_errors(),
        };

        let r = self
            .functions
            .borrow_mut()
            .entry(name.to_string())
            .insert_entry(function.into())
            .get()
            .clone();
        Some(r)
    }

    /// search in 'def's
    ///
    /// if `prefix` is found among the module's 'def's,
    /// this uses the registry to try to loading
    pub fn retrieve_module(
        &self,
        prefix: &str,
        registry: &ModuleRegistry,
    ) -> Option<IoResult<Rc<Module>>> {
        let path = &self
            .top
            .uses
            .iter()
            .rev()
            .find(|u| prefix == u.prefix)?
            .path;
        Some(registry.load(path))
    }

    fn compute_line_map(bytes: &[u8]) -> Box<[Range<usize>]> {
        let mut r = Vec::new();

        let mut it = bytes.iter();
        let mut at = 0;
        while let Some(nl) = it.position(|c| b'\n' == *c) {
            r.push(at..at + nl);
            at += nl + 1;
        }
        if at < bytes.len() {
            r.push(at..bytes.len());
        }

        r.into_boxed_slice()
    }

    pub fn get_containing_lnum(&self, k: usize) -> Option<usize> {
        self.line_map
            .iter()
            .position(|r| r.contains(&k))
            .map(|l| l + 1)
    }

    /// the tuple's first item is the same as if `source.get_containing_lnum(k.start)`
    pub fn get_containing_lines(&self, k: &Range<usize>) -> Option<(usize, &[Range<usize>])> {
        let mut it = self.line_map.iter();
        let start = it.position(|r| k.start <= r.end)?;
        let eof = self.line_map.len() - start;
        let end = start + it.position(|r| k.end < r.start).unwrap_or(eof - 1);
        Some((start + 1, &self.line_map[start..=end]))
    }
}
