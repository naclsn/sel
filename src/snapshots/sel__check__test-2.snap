---
source: src/check.rs
expression: "t(b\"{1, 2, 3}\")"
---
(
    Tree {
        ty: List(
            false,
            Named(
                "a",
                RefCell {
                    value: Some(
                        Number,
                    ),
                },
            ),
        ),
        val: Apply(
            Tree {
                ty: Func(
                    List(
                        false,
                        Named(
                            "a",
                            RefCell {
                                value: Some(
                                    Number,
                                ),
                            },
                        ),
                    ),
                    Func(
                        Named(
                            "a",
                            RefCell {
                                value: Some(
                                    Number,
                                ),
                            },
                        ),
                        List(
                            false,
                            Named(
                                "a",
                                RefCell {
                                    value: Some(
                                        Number,
                                    ),
                                },
                            ),
                        ),
                    ),
                ),
                val: Word(
                    "snoc",
                    Fundamental(
                        Snoc,
                    ),
                ),
            },
            [
                Tree {
                    ty: List(
                        false,
                        Named(
                            "a",
                            RefCell {
                                value: Some(
                                    Number,
                                ),
                            },
                        ),
                    ),
                    val: Apply(
                        Tree {
                            ty: Func(
                                List(
                                    false,
                                    Named(
                                        "a",
                                        RefCell {
                                            value: Some(
                                                Number,
                                            ),
                                        },
                                    ),
                                ),
                                Func(
                                    Named(
                                        "a",
                                        RefCell {
                                            value: Some(
                                                Number,
                                            ),
                                        },
                                    ),
                                    List(
                                        false,
                                        Named(
                                            "a",
                                            RefCell {
                                                value: Some(
                                                    Number,
                                                ),
                                            },
                                        ),
                                    ),
                                ),
                            ),
                            val: Word(
                                "snoc",
                                Fundamental(
                                    Snoc,
                                ),
                            ),
                        },
                        [
                            Tree {
                                ty: List(
                                    false,
                                    Named(
                                        "a",
                                        RefCell {
                                            value: Some(
                                                Named(
                                                    "a=item",
                                                    RefCell {
                                                        value: Some(
                                                            Number,
                                                        ),
                                                    },
                                                ),
                                            ),
                                        },
                                    ),
                                ),
                                val: Apply(
                                    Tree {
                                        ty: Func(
                                            List(
                                                false,
                                                Named(
                                                    "a",
                                                    RefCell {
                                                        value: Some(
                                                            Named(
                                                                "a=item",
                                                                RefCell {
                                                                    value: Some(
                                                                        Number,
                                                                    ),
                                                                },
                                                            ),
                                                        ),
                                                    },
                                                ),
                                            ),
                                            Func(
                                                Named(
                                                    "a",
                                                    RefCell {
                                                        value: Some(
                                                            Named(
                                                                "a=item",
                                                                RefCell {
                                                                    value: Some(
                                                                        Number,
                                                                    ),
                                                                },
                                                            ),
                                                        ),
                                                    },
                                                ),
                                                List(
                                                    false,
                                                    Named(
                                                        "a",
                                                        RefCell {
                                                            value: Some(
                                                                Named(
                                                                    "a=item",
                                                                    RefCell {
                                                                        value: Some(
                                                                            Number,
                                                                        ),
                                                                    },
                                                                ),
                                                            ),
                                                        },
                                                    ),
                                                ),
                                            ),
                                        ),
                                        val: Word(
                                            "snoc",
                                            Fundamental(
                                                Snoc,
                                            ),
                                        ),
                                    },
                                    [
                                        Tree {
                                            ty: List(
                                                true,
                                                Named(
                                                    "item",
                                                    RefCell {
                                                        value: Some(
                                                            Named(
                                                                "a=item",
                                                                RefCell {
                                                                    value: Some(
                                                                        Number,
                                                                    ),
                                                                },
                                                            ),
                                                        ),
                                                    },
                                                ),
                                            ),
                                            val: List,
                                        },
                                        Tree {
                                            ty: Number,
                                            val: Number(
                                                3.0,
                                            ),
                                        },
                                    ],
                                ),
                            },
                            Tree {
                                ty: Number,
                                val: Number(
                                    2.0,
                                ),
                            },
                        ],
                    ),
                },
                Tree {
                    ty: Number,
                    val: Number(
                        1.0,
                    ),
                },
            ],
        ),
    },
    [],
)
